### PR TITLE
Enable JSON Logging

### PR DESCRIPTION
**Pull request by erlanggaelfallujah**
Modified jpos log to support json format logging. Case if we shipping the logs file to a log aggregator and using json for standar message passing.

Below add format configuration (default format is xml) :

```
<log-listener class="org.jpos.util.DailyLogListener">
        <property name="format" value="json" />
        <property name="window" value="86400" />
        <property name="prefix" value="log/q2-json" />
        <property name="suffix" value=".log"/>
</log-listener>
```

Example Logs :

```
{
  "log": {
    "realm": "",
    "at": "2019-10-25T09:57:01.382233",
    "unpack": {
      "bitmap": "{3, 11, 24, 41, 46, 47, 48, 61}",
      "unpack": [
        {
          "packager": "org.jpos.iso.IFB_NUMERIC",
          "value": 450000,
          "fld": 3
        },
        {
          "packager": "org.jpos.iso.IFB_LLLBINARY",
          "value": {
            "type": "binary",
            "content": "DF90030"
          },
          "fld": 46
        }
      ]
    }
  }
}
{
  "log": {
    "realm": "",
    "at": "2019-10-25T09:57:01.377541",
    "error": "com.ranggalabs.swc.exception.Exception: UnLog at org.jpos.q2.iso.MUXPool.request(MUXPool.java:79) at com.ranggalabs.swc.Switcher.SessionManagerBean.sessionManaging(SessionManagerBean.java:80) "
  }
}
{
  "log": {
    "realm": "",
    "at": "2019-10-25T09:57:01.351143",
    "unpack": {
      "bitmap": "{3, 11, 24, 41, 46, 47, 48, 61}",
      "unpack": [
        {
          "packager": "org.jpos.iso.IFB_NUMERIC",
          "value": 450000,
          "fld": 3
        },
        {
          "packager": "org.jpos.iso.IFB_LLLBINARY",
          "value": {
            "type": "binary",
            "content": "DF90030"
          },
          "fld": 46
        }
      ]
    }
  }
}
{
  "log": {
    "realm": "",
    "at": "2019-10-25T09:57:01.361113",
    "lifespan": "4 ms",
    "receive": {
      "exception": {
        "sqlexception": "null",
        "sqlstate": "null",
        "vendorerror": "0",
        "stacktrace": [
          "java.sql.SQLException",
          "at org.jpos.util.JsonRotateListenerTest.testAddMessageAndSqlException(JsonRotateListenerTest.java:147)",
          "at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)",
          "at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)",
          "at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)",
          "at java.base/java.lang.reflect.Method.invoke(Method.java:567)",
          "at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)",
          "at java.base/java.lang.Thread.run(Thread.java:835)"
        ]
      }
    }
  }
}
{
  "log": {
    "realm": "",
    "at": "2019-10-25T09:57:01.308615",
    "info": {
      "isomsg": {
        "field": [
          {
            "id": 0,
            "value": "0800"
          },
          {
            "id": 3,
            "value": 990000
          },
          {
            "id": 7,
            "value": 101915
          }
        ],
        "header": "49534F303",
        "direction": "outgoing"
      }
    }
  }
}
```

### CODE REVIEW

**Comment by andresantoniuk** [comment]
I tried something similar in https://github.com/jpos/jPOS/pull/178 but I didn't manage to invest more time on that; hope this move forward as it's a really important improvement!

**Comment by mikesoares** [comment]
This is great! A few people from our team here at Zero Financial are currently looking this over.

**Comment by ar** [comment]
Great work. I'll review in detail, but have a few concerns:

- I fear this would break existing custom log listeners and filters (may be not)
- This substantial PR may have IP related issues (no copyright, no CLA nor CCLA)
- I see some internal company information leaked into the code in some tests
- Some System.out.printlns here and there that we try not to have in jPOS
- Some internal heavy used methods (such as log indenting) now require massive object creation and iterations on StringBuilders. I fear that code like this could have some performance issues, we are into the business of improving performance, not making it slower.
- Adds new dependency on old JSON libraries. We have the JSONPackager on hold until as a community decide which new JSON dependency we plan to use.

The jPOS Logger subsystem (I like to call it 'Audit subsystem' because it's not the same thing as traditional loggers such as log4j/j.u.l/slf4j/logback) is looking for some deep changes and improvements. Adding this PR at this time would perpetuate some of the faults we have.

But it is a great PR that raises the priority of deeper changes in the Logger subsystem, that I hope we can add during 2.1.5.

File: jpos/src/main/java/org/jpos/util/log/event/JSONLogEvent.java [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+                sb.append (" ms\"");
+            }
+
+            p.print(sb.toString());
+        }
+        return indent;
+    }
+
+    @Override
+    public void dumpTrailer(PrintStream p, String indent, boolean noArmor) {
```
Comment (richard-li-zero) [inline jpos/src/main/java/org/jpos/util/log/event/JSONLogEvent.java]: 'indent' is not used in this method

File: jpos/src/main/java/org/jpos/util/log/event/XMLLogEvent.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                sb.append ("ms\"");
+            }
+            sb.append ('>');
+            p.println (sb.toString());
+        }
+        return indent + "  ";
+    }
+
+    @Override
+    public void dumpTrailer(PrintStream p, String indent, boolean noArmor) {
```
Comment (richard-li-zero) [inline jpos/src/main/java/org/jpos/util/log/event/XMLLogEvent.java]: same as above.

File: jpos/src/main/java/org/jpos/util/log/event/JSONLogEvent.java (hunk 1 above)
Comment (erlanggaelfallujah) [reply jpos/src/main/java/org/jpos/util/log/event/JSONLogEvent.java]: updated, now is used.

File: jpos/src/main/java/org/jpos/util/log/event/XMLLogEvent.java (hunk 2 above)
Comment (erlanggaelfallujah) [reply jpos/src/main/java/org/jpos/util/log/event/XMLLogEvent.java]: It's legacy code, but it's used 

`p.println (indent + "</log>");`

**Comment by erlanggaelfallujah** [comment]
I was improve on indent process

**Comment by erlanggaelfallujah** [comment]
Hi @ar ,

- **I fear this would break existing custom log listeners and filters (may be not)**
Yeah, i think we need another case testing. And maybe, my unit test not covered this case.

- **This substantial PR may have IP related issues (no copyright, no CLA nor CCLA)**
Please guide me for this issues, because this first contribution to open source project.

- **Some internal heavy used methods (such as log indenting) now require massive object creation and iterations on StringBuilders. I fear that code like this could have some performance issues, we are into the business of improving performance, not making it slower.**
I was improved log indenting process, and need review again for other issues.

- **Adds new dependency on old JSON libraries. We have the JSONPackager on hold until as a community decide which new JSON dependency we plan to use.**
I was tried two dependencies (org.json and com.fasterxml.jackson). Jackson dependency have limitation, one of them the case that i found. If tag has duplicate keys they will get overwritten. So for alternative i am choose using org.json.

**Comment by ar** [comment]
I really appreciate the effort you put in this PR, but I believe this same goal can be achieved in a less intrusive way, implementing it as just a LogListener, perhaps added to the jPOS-EE project where different teams can start using the functionality, testing it performance, fixing problems, adding features, and then once battle tested perhaps graduate it to core jPOS. That's the way we usually do with most stuff. I suggest you follow-up the discussion in the jPOS slack, or the jPOS users mailing list. That said, I'll keep this PR open and perhaps we can merge it after 2.1.4 if the other route (log listener) is not feasible.

**Comment by ar** [comment]
Thank you for this contribution. closing it but we'll keep it in mind as a reference.

The jPOS Logger/Audit subsystem helped us well for over 20 years. Things are changing now, we are aware of jPOS systems running in over 500 nodes that generate logs that need to be aggregated. I don't think an adapter from the verbose XML logs into a JSON representation is the way to go. We hope to be able to provide a deeper and new approach to this problem.

**Pull request closed without merging**